File: DataEngineering/05_Stream_Processing/projects/python/01_realtime_dashboard_backend/main.py
import asyncio
import json
import logging
from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from kafka import KafkaConsumer
from typing import List

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = FastAPI(title="Real-Time Dashboard Backend")

# Configuration
KAFKA_BROKER = "localhost:9092"
KAFKA_TOPIC = "dashboard_metrics"

class ConnectionManager:
    def __init__(self):
        self.active_connections: List[WebSocket] = []

    async def connect(self, websocket: WebSocket):
        await websocket.accept()
        self.active_connections.append(websocket)
        logger.info("New WebSocket connection established")

    def disconnect(self, websocket: WebSocket):
        self.active_connections.remove(websocket)
        logger.info("WebSocket connection closed")

    async def broadcast(self, message: str):
        for connection in self.active_connections:
            try:
                await connection.send_text(message)
            except Exception as e:
                logger.error(f"Error sending message to client: {e}")

manager = ConnectionManager()

def get_kafka_consumer():
    """Initialize and return a Kafka consumer."""
    try:
        consumer = KafkaConsumer(
            KAFKA_TOPIC,
            bootstrap_servers=[KAFKA_BROKER],
            auto_offset_reset='latest',
            enable_auto_commit=True,
            group_id='dashboard-backend-group',
            value_deserializer=lambda x: x.decode('utf-8')
        )
        return consumer
    except Exception as e:
        logger.error(f"Failed to connect to Kafka: {e}")
        return None

async def consume_kafka_messages():
    """Background task to consume messages from Kafka and broadcast them to WS clients."""
    loop = asyncio.get_event_loop()
    consumer = await loop.run_in_executor(None, get_kafka_consumer)
    
    if not consumer:
        logger.warning("Kafka consumer not initialized. Skipping message consumption.")
        return

    logger.info("Started Kafka consumer background task")
    try:
        # We use a non-blocking loop pattern for demonstration
        for message in consumer:
            data = message.value
            logger.debug(f"Received message from Kafka: {data}")
            await manager.broadcast(data)
            # Yield control back to event loop
            await asyncio.sleep(0.01)
    except Exception as e:
        logger.error(f"Kafka consumption error: {e}")
    finally:
        consumer.close()

@app.on_event("startup")
async def startup_event():
    """Start background tasks on application startup."""
    asyncio.create_task(consume_kafka_messages())

@app.websocket("/ws/dashboard")
async def websocket_endpoint(websocket: WebSocket):
    """WebSocket endpoint for the real-time dashboard clients."""
    await manager.connect(websocket)
    try:
        while True:
            # We just wait for disconnects, communication is strictly server -> client
            data = await websocket.receive_text()
    except WebSocketDisconnect:
        manager.disconnect(websocket)

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)

File: DataEngineering/05_Stream_Processing/projects/golang/01_realtime_dashboard_backend/main.go
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
	"github.com/segmentio/kafka-go"
)

var (
	kafkaBroker = "localhost:9092"
	kafkaTopic  = "dashboard_metrics"
	upgrader    = websocket.Upgrader{
		CheckOrigin: func(r *http.Request) bool {
			return true // Allow all origins for demonstration
		},
	}
)

type Hub struct {
	clients    map[*websocket.Conn]bool
	broadcast  chan []byte
	register   chan *websocket.Conn
	unregister chan *websocket.Conn
	mu         sync.Mutex
}

func newHub() *Hub {
	return &Hub{
		broadcast:  make(chan []byte),
		register:   make(chan *websocket.Conn),
		unregister: make(chan *websocket.Conn),
		clients:    make(map[*websocket.Conn]bool),
	}
}

func (h *Hub) run() {
	for {
		select {
		case client := <-h.register:
			h.mu.Lock()
			h.clients[client] = true
			h.mu.Unlock()
			log.Println("New WebSocket client connected")
		case client := <-h.unregister:
			h.mu.Lock()
			if _, ok := h.clients[client]; ok {
				delete(h.clients, client)
				client.Close()
			}
			h.mu.Unlock()
			log.Println("WebSocket client disconnected")
		case message := <-h.broadcast:
			h.mu.Lock()
			for client := range h.clients {
				err := client.WriteMessage(websocket.TextMessage, message)
				if err != nil {
					log.Printf("Error writing to client: %v", err)
					client.Close()
					delete(h.clients, client)
				}
			}
			h.mu.Unlock()
		}
	}
}

func serveWs(h *Hub, w http.ResponseWriter, r *http.Request) {
	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Println("WebSocket upgrade error:", err)
		return
	}
	h.register <- conn

	// Keep connection alive and handle client disconnects
	go func() {
		defer func() {
			h.unregister <- conn
		}()
		for {
			_, _, err := conn.ReadMessage()
			if err != nil {
				break
			}
		}
	}()
}

func consumeKafkaMessages(ctx context.Context, h *Hub) {
	r := kafka.NewReader(kafka.ReaderConfig{
		Brokers:  []string{kafkaBroker},
		Topic:    kafkaTopic,
		GroupID:  "dashboard-backend-group",
		MinBytes: 10e3, // 10KB
		MaxBytes: 10e6, // 10MB
	})
	defer r.Close()

	log.Printf("Started consuming Kafka topic: %s", kafkaTopic)

	for {
		m, err := r.ReadMessage(ctx)
		if err != nil {
			log.Printf("Kafka read error: %v", err)
			time.Sleep(2 * time.Second) // backoff
			continue
		}
		// Broadcast to all connected clients
		h.broadcast <- m.Value
	}
}

func main() {
	fmt.Println("Starting RealTime Dashboard Backend...")

	hub := newHub()
	go hub.run()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Start Kafka consumer in the background
	go consumeKafkaMessages(ctx, hub)

	http.HandleFunc("/ws/dashboard", func(w http.ResponseWriter, r *http.Request) {
		serveWs(hub, w, r)
	})

	addr := ":8080"
	log.Printf("Server starting on %s", addr)
	if err := http.ListenAndServe(addr, nil); err != nil {
		log.Fatalf("Server error: %v", err)
	}
}
